content/blog/2017/05/2017-05-12-kubernetes-journey-on-azure.adoc

```diff
@@ -0,0 +1,397 @@
+      }
+    }
+    stage('Deploy'){
+      steps{
+       // Deploy changes in production
+       sh 'make deploy'
+      }
+    }
+  }
+  post{
+    always {
+      sh 'make notify'
+    }
+  }
+}
+----
+
+=== Secret credentials
+
+Managing secrets is a big subject and brings with it many different
+requirements which are very hard to fulfill.  For obvious reasons, we couldn't
+publish the credentials used within the infra project.  On the other side, we
+needed to keep track and share them. Particularly for the Jenkins node that
+deploys our cluster.  This means that we needed a way to encrypt or decrypt
+those credentials depending on permissions, environments, etc.  We analyzed two
+different approaches to handle this:
+
+  1. Storing secrets in a key management tool like https://azure.microsoft.com/en-us/services/key-vault/[Key Vault] or https://www.vaultproject.io/[Vault] and use them like a kubernetes "secret" type of resource. +
+    -> Unfortunately, these tools are not yet integrated in Kubernetes. But we may come back to this option later.
+    https://github.com/kubernetes/kubernetes/issues/10439[Kubernetes issue: 10439]
+
+  2. Publishing and encrypting using a public gpg key. +
+     This means that everybody can encrypt credentials for the infrastructure project but only the owner of the private key can decrypt credentials. +
+     This solution implies
+      * Scripting, as secrets need to be decrypted at deployment time.
+      * Templates, as secrets value will change depending on the environment. +
+     -> Each Jenkins node should only have the private key to decrypt secrets associated to its environment.
+
+=== Scripting
+
+Finally, the system we had built was hard to work with.  Our initial
+`Jenkinsfile`, with only one `kubectl` command to run, became a bunch of
+scripts. There were so many situations requiring additional steps.
+
+* Resources needed to be updated only in some situations
+* Secrets needed to be encrypted/decrypted
+* Tests needed to be run.
+
+At the end, the amount of scripts used to deploy the kubernetes resources
+started to grow a lot.  And we started asking ourselves: "Aren't we
+reinventing the wheel?"
+
+== The Puppet way
+
+The Jenkins project already uses Puppet, so we decided to look at using Puppet
+to orchestrate our container deployment with Kubernetes.
+
+.Workflow
+----
+  _________________       ____________________       _____________
+  |                |      |                   |      |            |
+  |    Github:     |      |     Jenkins:      |      | Puppet:    |
+  | Commit trigger | ---->| Test & Validation | ---->| Deployment |
+  |________________|      |___________________|      |____________|
+----
+
+
+In this workflow, Puppet is used to template and deploy all kubernetes
+configurations files needed to orchestrate our cluster in a controlled
+workspace. It is also used to automate basic operations like 'apply' or
+'remove' resources based on file changed.
```

Comment:
Sentence looks broken.

* file changes?
* A comma or two missing?
* Are 'apply' and 'remove' resources, or how is this meant?